net/dns: TestWatchFile is flaky and timing out in CI

Examples:
https://github.com/tailscale/tailscale/actions/runs/17769808669/job/50502582943
https://github.com/tailscale/tailscale/actions/runs/17695675344/job/50516603830
https://github.com/tailscale/tailscale/actions/runs/17684049459/job/50264664075
https://github.com/tailscale/tailscale/actions/runs/17622308002/job/50070638754

```
=== RUN   TestWatchFile
panic: test timed out after 10m0s
	running tests:
		TestWatchFile (10m0s)
```

--- a/net/dns/direct_linux_test.go
+++ b/net/dns/direct_linux_test.go
@@ -25,8 +25,13 @@ func TestWatchFile(t *testing.T) {
 	var callbackCalled atomic.Bool
 	callbackDone := make(chan bool)
 	callback := func() {
-		callbackDone <- true
-		callbackCalled.Store(true)
+		// We only send to the channel once to avoid blocking if the
+		// callback is called multiple times -- this happens occasionally
+		// if inotify sends multiple events before we cancel the context.
+		if !callbackCalled.Load() {
+			callbackDone <- true
+			callbackCalled.Store(true)
+		}
 	}
 
 	var eg errgroup.Group

